File: tutorials/019_consume_rtsp_stream/consume_rtsp_stream.cpp
// Consume a live H.264 RTSP stream via the RtspDecodedInput node group.
//
// The group handles RTSP connect, depacketize, and H.264 decode — you hand it
// a URL and pull decoded frames. This chapter is about the input group only.
//
// Usage:
//   tutorial_v2_019_consume_rtsp_stream --url rtsp://host/path [--frames 5]

#include "neat.h"
#include "nodes/groups/RtspDecodedInput.h"

#include <exception>
#include <iostream>
#include <stdexcept>
#include <string>

namespace {

bool get_arg(int argc, char** argv, const std::string& key, std::string& out) {
  for (int i = 1; i + 1 < argc; ++i) {
    if (key == argv[i]) {
      out = argv[i + 1];
      return true;
    }
  }
  return false;
}

int parse_int_arg(int argc, char** argv, const std::string& key, int def) {
  std::string v;
  if (!get_arg(argc, argv, key, v))
    return def;
  return std::stoi(v);
}

} // namespace

int main(int argc, char** argv) {
  try {
    std::string url;
    if (!get_arg(argc, argv, "--url", url)) {
      std::cerr << "Usage: tutorial_v2_019_consume_rtsp_stream --url <rtsp://...> [--frames <n>]\n";
      return 1;
    }
    const int frames = parse_int_arg(argc, argv, "--frames", 5);

    // CORE LOGIC
    // Configure RtspDecodedInputOptions, build a Session whose only stages
    // are the RTSP group and an Output node, and pull decoded frames.
    simaai::neat::nodes::groups::RtspDecodedInputOptions rtsp_opt;
    rtsp_opt.url = url;
    rtsp_opt.tcp = true;

    simaai::neat::Session s;
    s.add(simaai::neat::nodes::groups::RtspDecodedInput(rtsp_opt));
    s.add(simaai::neat::nodes::Output());
    auto run = s.build(simaai::neat::RunOptions{});

    for (int i = 0; i < frames; ++i) {
      auto sample = run.pull(/*timeout_ms=*/5000);
      if (!sample.has_value() || !sample->tensor.has_value()) {
        std::cout << "frame=" << i << " rtsp_timeout\n";
        break;
      }
      const auto& shape = sample->tensor->shape;
      std::cout << "frame=" << i << " shape=[";
      for (std::size_t d = 0; d < shape.size(); ++d) {
        std::cout << shape[d] << (d + 1 < shape.size() ? ", " : "");
      }
      std::cout << "]\n";
    }
    // END CORE LOGIC

    return 0;
  } catch (const std::exception& e) {
    std::cerr << "[FAIL] " << e.what() << "\n";
    return 1;
  }
}

File: tutorials/017_consume_rtsp_stream/consume_rtsp_stream.py
#!/usr/bin/env python3
"""Consume a live H.264 RTSP stream via the RtspDecodedInput Graph fragment.

The fragment handles RTSP connect, depacketize, and H.264 decode — you hand it a
URL and pull decoded frames. This chapter is about the input fragment only.

Usage:
  python3 consume_rtsp_stream.py --url rtsp://host/path [--frames 5]
"""
from __future__ import annotations

import argparse
import sys

try:
  import pyneat
except ImportError:
  sys.exit(
      "pyneat is not importable. Either Neat is not installed, or the venv is not activated.\n"
      "Run: source ~/pyneat/bin/activate"
  )


def main(argv: list[str]) -> int:
  ap = argparse.ArgumentParser(description=__doc__)
  ap.add_argument("--url", required=True, help="RTSP URL to consume")
  ap.add_argument("--frames", type=int, default=5, help="Frames to pull")
  args = ap.parse_args(argv[1:])

  # CORE LOGIC
  # Configure RtspDecodedInputOptions, build a Graph whose only stages are
  # the RTSP group and an output node, and pull decoded frames.
  rtsp_opt = pyneat.RtspDecodedInputOptions()
  rtsp_opt.url = args.url
  rtsp_opt.tcp = True

  s = pyneat.Graph()
  s.add(pyneat.groups.rtsp_decoded_input(rtsp_opt))
  s.add(pyneat.nodes.output())
  run = s.build(pyneat.RunOptions())

  for i in range(args.frames):
    sample = run.pull(timeout_ms=5000)
    if sample is None or sample.tensor is None:
      print(f"frame={i} rtsp_timeout")
      break
    print(f"frame={i} shape={list(sample.tensor.shape)}")
  # END CORE LOGIC
  return 0


if __name__ == "__main__":
  raise SystemExit(main(sys.argv))
